core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp

```diff
@@ -408,14 +407,15 @@ class TeamPolicyInternal<Kokkos::Cuda, Properties...>
 };
 
 __device__ inline int64_t cuda_get_scratch_index(Cuda::size_type league_size,
-                                                 int32_t* scratch_locks) {
+                                                 int32_t* scratch_locks,
+                                                 size_t num_scratch_locks) {
   int64_t threadid = 0;
   __shared__ int64_t base_thread_id;
   if (threadIdx.x == 0 && threadIdx.y == 0) {
     int64_t const wraparound_len = Kokkos::max(
-        int64_t(1), Kokkos::min(int64_t(league_size),
-                                (int64_t(g_device_cuda_lock_arrays.n)) /
-                                    (blockDim.x * blockDim.y)));
+        int64_t(1),
+        Kokkos::min(int64_t(league_size),
+                    int64_t(num_scratch_locks) / (blockDim.x * blockDim.y)));
```

Comment:
I think this is wrong. You defined `num_scratch_locks` as `sizeof(int32_t) * concurrency` but `g_device_cuda_lock_arrays.n` is equal to `concurrency`.

Reply:
Amending
```diff
diff --git a/core/src/Cuda/Kokkos_Cuda_Instance.cpp b/core/src/Cuda/Kokkos_Cuda_Instance.cpp
index f0816ea44..45fdbd641 100644
--- a/core/src/Cuda/Kokkos_Cuda_Instance.cpp
+++ b/core/src/Cuda/Kokkos_Cuda_Instance.cpp
@@ -421,10 +421,11 @@ Kokkos::Cuda::initialize WARNING: Cuda is allocating into UVMSpace by default
     m_team_scratch_ptr[i]          = nullptr;
   }

-  m_num_scratch_locks = sizeof(int32_t) * concurrency();
-  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMalloc(&m_scratch_locks, m_num_scratch_locks));
+  m_num_scratch_locks = concurrency();
   KOKKOS_IMPL_CUDA_SAFE_CALL(
-      cudaMemset(m_scratch_locks, 0, m_num_scratch_locks));
+      cudaMalloc(&m_scratch_locks, sizeof(int32_t) * m_num_scratch_locks));
+  KOKKOS_IMPL_CUDA_SAFE_CALL(
+      cudaMemset(m_scratch_locks, 0, sizeof(int32_t) * m_num_scratch_locks));
 }

 //----------------------------------------------------------------------------
diff --git a/core/src/HIP/Kokkos_HIP_Instance.cpp b/core/src/HIP/Kokkos_HIP_Instance.cpp
index 4ffef2b99..7840ad905 100644
--- a/core/src/HIP/Kokkos_HIP_Instance.cpp
+++ b/core/src/HIP/Kokkos_HIP_Instance.cpp
@@ -180,9 +180,11 @@ void HIPInternal::initialize(hipStream_t stream, bool manage_stream) {
     Kokkos::Impl::throw_runtime_exception(msg.str());
   }

-  m_num_scratch_locks = sizeof(int32_t) * concurrency();
-  KOKKOS_IMPL_HIP_SAFE_CALL(hipMalloc(&m_scratch_locks, m_num_scratch_locks));
-  KOKKOS_IMPL_HIP_SAFE_CALL(hipMemset(m_scratch_locks, 0, m_num_scratch_locks));
+  m_num_scratch_locks = concurrency();
+  KOKKOS_IMPL_HIP_SAFE_CALL(
+      hipMalloc(&m_scratch_locks, sizeof(int32_t) * m_num_scratch_locks));
+  KOKKOS_IMPL_HIP_SAFE_CALL(
+      hipMemset(m_scratch_locks, 0, sizeof(int32_t) * m_num_scratch_locks));
 }

 //----------------------------------------------------------------------------